keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1054 @@
+  to ensure that configuration parameters are available when the DRA driver
+  needs them.
+
+The types that the scheduler needs to watch may change at runtime as classes
+get added or removed. To avoid repeated GET operations for those CRDs, dynamic
+informers will have to be used that get started and stopped as needed.
+
+## Design Details
+
+### ResourceClass extension
+
+A new, optional field in ResourceClass enables semantic parameters for claims
+using this class:
+
+```
+type ResourceClass struct {
+	...
+
+	// If (and only if) allocation of claims using this class is handled
+	// via numeric parameters, then ParameterTypes describes those
+	// parameters. A driver might support multiple different CRDs for
+	// its claims.
+	ParameterTypes []SemanticParameterType
+}
+
+// SemanticParameterType describes one resource that may be used for
+// claim parameters of a class which uses semantic parameters.
+type SemanticParameterType struct {
+	// APIGroup is the group for the resource being referenced. It is
+	// empty for the core API. This matches the group in the APIVersion
+	// that is used when creating the resources.
+	// +optional
+	APIGroup string
+	// Kind is the type of resource being referenced. This is the same
+	// value as in the parameter object's metadata.
+	Kind string
+
+	// Shareable is copied into the AllocationResult when a claim
+	// gets allocated.
+	Shareable bool
+}
+```
+
+### NodeResources
+
+For each node, one or more NodeResources objects gets created:
+
+```
+// NodeResources provides information about available
+// resources on individual nodes.
+type NodeResources struct {
+	metav1.TypeMeta
+	// Standard object metadata
+	// +optional
+	metav1.ObjectMeta
+
+	// NodeName identifies the node where the capacity is available.
+	// A field selector can be used to list only NodeResources
+	// objects with a certain node name.
+	NodeName string
+
+	// DriverName identifies the DRA driver providing the capacity information.
+	// A field selector can be used to list only NodeResources
+	// objects with a certain driver name.
+	DriverName string
+
+	// Instances describes discrete resources managed by the driver on
+	// the node. It is possible to create more than one NodeResources
+	// object for the same node and driver if this array becomes to large
+	// for a single object.
+	Instances []NodeResourceInstance
+}
+
+// NodeResourceInstance describes one discrete resource instance.
+type NodeResourceInstance struct {
+	// ID is chosen by the driver to distinguish between different resource
+	// instances. It only needs to be unique for the driver and node.  In
+	// other words, the tuple of NodeName, DriverName, InstanceID needs to
+	// be unique in the cluster.
+	ID string
+
+    // Attributes describe the instance.
+	Attributes []InstanceAttributes
+}
+
+// InstanceAttributes must have one and only one field set.
+type InstanceAttributes struct {
+	<one pointer per semantic model>
+}
+```
```

Comment:
The "parameters" side of this seems to be missing. You have instance attributes defined. But there should be a common ClaimParameter model too, which basically has an array of union types. You sort of define it above but not here in the details section, which is a little confusion.

I think the detailed design section should follow basically the same flow you put up in the summary, but add all the basic API objects your talking about, and sort of explain how each is created and managed.

Now, I think we can take it farther. In particular I think for the selection, we may want to consider a CEL expression. So, if the resource attributes for selection are basically a bunch of key-value pairs, and we make them essentially universal (with driver-specific ones allowed), then you can bump the selection expression up to the ResourceClass (and maybe tighter selection in the ResourceClaim). So, you might have something like (not sure the right apiVersions here):


```yaml
apiVersion: core.k8s.io/v1alpha3
kind: ResourceClass
metadata:
  name: acme1000-gpu-v3-plus
driverName: gpu.example.com
parametersRef:
  apiGroup: gpu.example.com
  kind: GPUInit
  name: acme1000-gpu-v3-plus
selection: "model == 'acme1000' && driverMajorVersion >= 3"
---
apiVersion: resource.k8s.io/v1alpha2
kind: ResourceClaim
metadata:
  name: myclaim
spec:
  resourceClassName: "acme-gpu"
  parametersRef:
    apiGroup: gpu.example.com
    kind: GPURequirements
    name: device-consumer-gpu-parameters
  selection: "driverMajorVersion == 3 && driverMinorVersion == 2 && driverPatchVersion >= 7 && attributes['gpu.example.com/foo'] == 'bar'"
```

Basically, we make the common things people need to do, like selection, common. We leave the CR to cover the really really specific stuff. In this case, the vendor can create different classes that select on top of the set of attributes, and the claim author can further refine that if they want to. The selection attributes can include well-known ones (model, driverVersion) as well as driver-specific ones (`gpu.example.com/foo`).

This prevents every vendor from having to create their boolean expressions, etc. to allow selection.

The user *already* has to write a resource claim as well as a vendor-specific claim parameter. So, I don't think adding this to the resource claim is a worse UX. In fact, I think it's better, because otherwise every vendor will have their own version of this.

Alternatively, this "selection" behavior could be stuffed into a semantic model itself. But I think that would be more confusing.

Reply:
> Basically, we make the common things people need to do, like selection, common.

I don't see "model", "driverMinorVersion", etc. as something that is common. Even if we were to pre-define them as fields or names in the resource.k8s.io API, their semantic would be mostly left to vendors. Let's start by defining selection via a semantic model. If we then find that it deserves to be a first-class citizen, we can do that as a future step.

I agree with making selection a CEL expression. We can make it type-safe by defining not just `labels map[string]string` where the value has to be interpreted at runtime, but instead `labels` (really just opqaque strings), `quantities` (values are `resource.Quantity`) and `versions` (values are semver).

I am not sure whether we need semver. It is more expressive than a (major, minor, patch) tuple. If such a tuple is enough, then it can be stored in `quantities` with three different names. Hmm, we probably want it. The boolean expression for `(major1, minor1, patch1) > (major2, minor2, patch2)` is complicated.

Reply:
> The "parameters" side of this seems to be missing.

Right. Looks like I left out the Go type for `ResourceClaimParameters` and only mentioned it in the intro. Will add.

Reply:
I can see how having a `selection` (or `selector`) field available at this level could be useful for drivers that don't want to define their own selection API. For those that do, however, it might be confusing to have this field available here (which one takes precedence if both are set, for example).

I would opt for not including this in the initial iteration, and instead create a builtin, generic `ClaimParameters` type against the `resource.k8s.io` API group that includes this. When referenced in a claim, the scheduler will know how to read / interpret these parameters directly without having to delegate their translation to a DRA driver controller.

Reply:
> useful for drivers that don't want to define their own selection API

At risk of repeating: IMO, selection (constraint resolution) is a generic process.  Devices offer information and users write CEL against that information.  We don't need to dictate what is IN that information, just the "query language" used to evaluate it.

Reply:
Where the values to match the selector against may depend on the semantic model that describes the resources. Not all models will have that information "at the root".

Reply:
The current "partitionable" model has attributes at all levels.